### PR TITLE
Introduce quarkus-jackson and quarkus-resteasy-jackson extensions

### LINKED ISSUE #3345: quarkus-resteasy-jackson needed
**Description**
Please provide a configuration dependency to use jackson as object mapper instead of json-b.
The reason for this is the superior support for Kotlin data classes in jackson. Using jackson-module-kotlin jackson can directly marshal data classes w/o a no args constructor.

### PR DESCRIPTION
**Pull request by geoand**
Besides adding the proper dependencies, this extension also makes it
easy for users to register a custom ObjectMapper by allowing them to
simply register a bean of that type

Fixes #3345

### CODE REVIEW

**Comment by gsmet** [comment]
Looks mostly good to me but I wonder if we should have a `quarkus-jackson` extension too, similar to what we have with JSON-B.

If people needs Jackson outside of RESTEasy, they will need the reflection registrations.

**Comment by geoand** [comment]
@gsmet I am going to add another commit in a few minutes to address a corner case.

Once I do that, then I can also create the separate extension for the reflection.

**Comment by geoand** [comment]
@gsmet I now also added a `quarkus-jackson` extension that simply contains the reflection registration

**Comment by machi1990** [comment]
Do we have to add the two extensions into the `extensions.json` file?

**Comment by geoand** [comment]
Ah yes, I forgot about that, thanks @machi1990.
I'll take care of it in a while

**Comment by geoand** [comment]
I will add the entries to the extension json file later on today because I need to step out for a few hours

File: extensions/resteasy-jackson/runtime/pom.xml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-jackson2-provider</artifactId>
+        </dependency>
```
Comment (gsmet) [inline extensions/resteasy-jackson/runtime/pom.xml]: You should have a dependency for `quarkus-jackson` here.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a couple more comments.

File: extensions/resteasy-jackson/deployment/pom.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jackson-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-jackson</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jackson</artifactId>
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-jackson/deployment/pom.xml]: Great catch, I mixed up the position of this dependency :)

File: extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                        instanceNull.load(ObjectMapper.class.getName()));
+                ResultHandle instanceHandle = instanceNull.invokeInterfaceMethod(
+                        MethodDescriptor.ofMethod(ArcContainer.class, "instance", InstanceHandle.class, Class.class,
+                                Annotation[].class),
+                        arcContainer, jsonbClass, instanceNull.loadNull());
+                ResultHandle get = instanceNull.invokeInterfaceMethod(
+                        MethodDescriptor.ofMethod(InstanceHandle.class, "get", Object.class),
+                        instanceHandle);
+
+                ResultHandle jsonb = instanceNull.checkCast(get, ObjectMapper.class);
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java]: Fixed

File: extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                continue;
+            }
+            if (OBJECT_MAPPER.equals(annotation.target().asMethod().returnType().name())) {
+                result.add(annotation.target().asMethod().declaringClass().name().toString());
+            }
+        }
+        return result;
+    }
+
+    private void generateObjectMapperContextResolver(ClassOutput classOutput) {
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java]: Hopefully it's clearer now

**Comment by geoand** [comment]
@gsmet comments addressed (hopefully :)). @machi1990 `extensions.json` file updated

File: extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+import io.quarkus.gizmo.ResultHandle;
+import io.quarkus.resteasy.common.deployment.ResteasyJaxrsProviderBuildItem;
+import io.quarkus.resteasy.jackson.runtime.ObjectMapperProducer;
+
+public class ResteasyJacksonProcessor {
+
+    private static final DotName OBJECT_MAPPER = DotName.createSimple(ObjectMapper.class.getName());
+    private static final DotName CONTEXT_RESOLVER = DotName.createSimple(ContextResolver.class.getName());
+
+    private static final String QUARKUS_CONTEXT_RESOLVER_NAME = "io.quarkus.jackson.QuarkusObjectMapperContextResolver";
```
Comment (gsmet) [inline extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java]: Shouldn't the package be `io.quarkus.resteasy.jackson.runtime`?

It's not API, thus the runtime. And it's related to RESTEasy.

**Review by gsmet — CHANGES_REQUESTED** [review]
@geoand I added one minor comment, could you check it out?

Otherwise, it looks good. But I think we should probably add some notes in the documentation mentioning that Jackson is also supported.

**Comment by geoand** [comment]
@gsmet sure thing, I'll take care of it tonight

File: extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java (hunk 5 above)
Comment (geoand) [reply extensions/resteasy-jackson/deployment/src/main/java/io/quarkus/resteasy/jackson/deployment/ResteasyJacksonProcessor.java]: Good idea, I like it

**Comment by geoand** [comment]
@gsmet I updated the PR according to your suggestions, does it look good now?

**Review by gsmet — APPROVED** [review]
Great work, thanks!

**Comment by geoand** [comment]
You are very welcome!

**Comment by cg-d** [comment]
Thanks @geoand. We will checkout 0.20.0 and give feedback :-)

**Comment by geoand** [comment]
Sounds good!

**Comment by cg-d** [comment]
Sorry for the late feedback. Jackson is working like a charm :+1:

**Pull request merged**